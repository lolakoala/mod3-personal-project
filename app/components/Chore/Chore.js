import React, { Component } from 'react';
import PropTypes from 'prop-types';

class Chore extends Component {
  getUserAssigned = () => {
    const { usersHouse } = this.props;
    const chore = usersHouse.chores.find(chore => chore.id === this.props.match.params.id);
    return usersHouse.users.find(user => user.id === chore.assignedTo).name;
  }

  claimChore = (userId, chore, usersHouse) => {
    this.props.claimChore(userId, chore, usersHouse);
    this.forceUpdate();
  }

  markChoreDone = (chore, usersHouse) => {
    this.props.markChoreDone(chore, usersHouse);
    this.forceUpdate();
  }

  render() {
    const { usersHouse, currentUser } = this.props;
    const chore = usersHouse.chores.find(chore => chore.id === this.props.match.params.id);
    return (
      <div>
        <p className='title'>{`Chore: ${chore.title}`}</p>
        <p className='urgency'>{`Urgency: ${chore.urgency}`}</p>
        <p className='posted-by-on'>{`Posted by ${chore.postedBy.name} on ${chore.datePosted}`}</p>
        <p className='claim-chore'
          onClick={() => this.claimChore(currentUser.id, chore, usersHouse)}>
          {chore.assignedTo.length ? `Claimed by ${this.getUserAssigned()}` : 'Claim Chore'}
        </p>
        <p className='details'>{`Details: ${chore.details}`}</p>
        <p className='chore-done' onClick={() => this.markChoreDone(chore, usersHouse)}>{chore.done === false ? 'Mark as Done' : 'Done'}</p>
      </div>
    );
  }

}

export default Chore;

Chore.propTypes = {
  usersHouse: PropTypes.object,
  match: PropTypes.object,
  markChoreDone: PropTypes.func,
  currentUser: PropTypes.object,
  claimChore: PropTypes.func
};
