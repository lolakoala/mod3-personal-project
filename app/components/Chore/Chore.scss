#chore-page {
  margin: 20px;
  height: auto;
  width: 280px;
  border-radius: 5px;
  background-color: rgba(141, 132, 73, 0.7);
  color: #3d6aa2;
  padding: 20px;
  font-size: 20px;
}

#chore-page .title {
  height: auto;
  width: 90%;
  background-color: #F4CCE2;
  border-radius: 5px;
  padding: 5px;
  text-align: center;
  font-size: 30px;
  margin-top: 10px;
}

#chore-page .details {
  height: auto;
  width: 90%;
  background-color: #F4CCE2;
  border-radius: 5px;
  padding: 5px;
  margin-top: 10px;
  margin-bottom: 10px;
}

.urgency-done,
#chore-page .claim-chore {
  height: auto;
  width: 90%;
  background-color: #F4CCE2;
  border-radius: 5px;
  margin-top: 10px;
  display: flex;
  padding: 5px;
  position: relative;
}

.urgency-done .chore-done {
  height: 100%;
  width: 85px;
  background-color: #eb6b3c;
  position: absolute;
  top: 0;
  right: 0;
  color: #F4CCE2;
  display: inline-block;
  font-size: 20px;
  text-align: center;
  border-radius: 5px;
  // padding-top: 5px;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
}

#chore-page .posted-by-on {
  display: inline-block;
  margin-top: 10px;
}

#chore-page .claim-chore {
  background-color: #eb6b3c;
  color: #F4CCE2;
  text-align: center;
}

.chore-done:hover,
#chore-page .claim-chore:hover {
  cursor: pointer;
  background-color: #F4CCE2;
  color: #eb6b3c;
}
