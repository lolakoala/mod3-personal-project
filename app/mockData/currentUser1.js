const currentUser1 = {
  id : "RVde1m9BhkQBLZlTJNsFlQZXqP43",
  name: "Amanda Brenner",
  house: "Dog9"
};

export default currentUser1;
