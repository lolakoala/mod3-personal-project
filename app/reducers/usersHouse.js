import firebase from '../firebase.js';

const usersHouse = (state = {}, action) => {
  const housesRef = firebase.database().ref('houses');
  switch (action.type) {
  case 'CREATE_HOUSE':
    housesRef.push(action.house);
    return action.house;
  case 'GET_HOUSE':
    const updatedHouse = Object.assign(action.usersHouse, { users: [...action.usersHouse.users, action.user] });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(updatedHouse);
    return updatedHouse;
  case 'LOGIN_SUCCESS':
    return action.usersHouse;
  case 'SIGNOUT':
    return {};
  case 'LEAVE_HOUSE':
    const deleteUser = action.usersHouse.users.filter(user => user.id !== action.currentUser.id);
    const houseWithoutUser = Object.assign(action.usersHouse, { users: deleteUser });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(houseWithoutUser);
    return {};
  case 'ADD_BILL':
    const takeOutFakeBills = action.usersHouse.bills.filter(bill => bill.title !== 'fake');
    const houseWithBill = Object.assign(action.usersHouse, { bills: [...takeOutFakeBills, action.bill] });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(houseWithBill);
    return houseWithBill;
  case 'ADD_BULLETIN':
    const takeOutFakeBulletins = action.usersHouse.bulletins.filter(bulletin => bulletin.title !== 'fake');
    const houseWithBulletin = Object.assign(action.usersHouse, { bulletins: [...takeOutFakeBulletins, action.bulletin] });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(houseWithBulletin);
    return houseWithBulletin;
  case 'ADD_CHORE':
    const takeOutFakeChore = action.usersHouse.chores.filter(chore => chore.title !== 'fake');
    const houseWithChore = Object.assign(action.usersHouse, { chores: [...takeOutFakeChore, action.chore] });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(houseWithChore);
    return houseWithChore;
  case 'ADD_READER':
    const oldBulletin = action.usersHouse.bulletins.find(bulletin => bulletin.id === action.bulletinId);
    if (oldBulletin.hasRead.includes(action.userId)) {
      return action.usersHouse;
    }
    const newBulletin = Object.assign(oldBulletin, { hasRead: [...oldBulletin.hasRead, action.userId] });
    const updateBulletins = action.usersHouse.bulletins.filter(bulletin => {
      return bulletin.id !== action.bulletinId;
    });
    const updateWithBulletin = Object.assign(action.usersHouse, { bulletins: [...updateBulletins, newBulletin] });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(updateWithBulletin);
    return updateWithBulletin;
  case 'MARK_PAID':
    const oldBill = action.usersHouse.bills.find(bill => bill.id === action.billId);
    const user = oldBill.allUsersTotals.find(user => user.id === action.userId);
    if (user.paid === true) {
      return action.usersHouse;
    }
    const newUser = Object.assign(user, { paid: true });
    const newAllUsersTotals = oldBill.allUsersTotals.filter(user => user.id !== action.userId);
    const newBill = Object.assign(oldBill, { allUsersTotals: [...newAllUsersTotals, newUser] });
    const updatedBills = action.usersHouse.bills.filter(bill => {
      return bill.id !== action.billId;
    });
    const updateWithBill = Object.assign(action.usersHouse, { bills: [...updatedBills, newBill] });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(updateWithBill);
    return updateWithBill;
  default:
    return state;
  }
};

export default usersHouse;
