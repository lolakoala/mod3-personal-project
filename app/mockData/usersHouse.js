import currentUser1 from './currentUser1.js';
import currentUser2 from './currentUser2.js';

const bill1 = {
  allUsersTotals: [
    {
      id: currentUser1.id,
      paid: false,
      total: 50
    },
    {
      id: currentUser2.id,
      paid: false,
      total: 50
    }
  ],
  datePosted: "11/01/2017",
  details: "We used a lot of water.",
  duedate: "11/9/17",
  id: "b3019762705779",
  parsedDuedate: 1510210800000,
  postedBy: currentUser1,
  title: "Water Bill",
  total: "100"
};

const bulletin1 = {
  datePosted: "11/01/2017",
  details: "The dogs got out through a hole in the fence yesterday.",
  hasRead: [currentUser1.id],
  id: 1509567098747,
  postedBy: currentUser1,
  title: "Don't leave dogs out unattended."
};

const chore1 = {
  assignedTo: currentUser1.id,
  datePosted: "10/31/2017",
  details: "Empty the dishwasher.",
  done: false,
  id: "c1509485241076",
  postedBy: currentUser1,
  title: "Dishes",
  urgency: "low"
};

const usersHouse = {
  houseKey: "-KxoIBWy5UYRscFYdWty",
  bills: [bill1],
  bulletins: [bulletin1],
  chores: [chore1],
  houseCode: "woof9",
  houseName: "Dog9",
  users: [currentUser1, currentUser2]
};

export default usersHouse;
