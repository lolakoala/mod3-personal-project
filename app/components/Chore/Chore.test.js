import React from "react";
import Chore from './Chore.js';
import { shallow } from "enzyme";
import usersHouse from '../../mockData/usersHouse.js';
import currentUser1 from '../../mockData/currentUser1.js';

describe('chore component', () => {
  const mockFn = jest.fn();
  const chore = usersHouse.chores[0];
  const match = {
    params: {
      id: "c1509485241076"
    }
  };
  const wrapper = shallow(<Chore
    usersHouse={usersHouse}
    currentUser={currentUser1}
    claimChore={mockFn}
    markChoreDone={mockFn}
    match={match}/>);

  it('should exist', () => {
    expect(wrapper).toBeDefined();
  });

  it('should have a title', () => {
    const title = wrapper.find('.title');

    expect(title.text()).toEqual(`Chore: ${chore.title}`);
  });
});
