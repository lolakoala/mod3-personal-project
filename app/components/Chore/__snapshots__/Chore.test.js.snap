// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chore component should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Chore
    claimChore={[Function]}
    currentUser={
      Object {
        "house": "Dog9",
        "id": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
        "name": "Amanda Brenner",
      }
    }
    markChoreDone={[Function]}
    match={
      Object {
        "params": Object {
          "id": "c1509485241076",
        },
      }
    }
    usersHouse={
      Object {
        "bills": Array [
          Object {
            "allUsersTotals": Array [
              Object {
                "id": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
                "paid": false,
                "total": 50,
              },
              Object {
                "id": "cFNdorl5ceRyMr9XGvPFRVEcw7I3",
                "paid": false,
                "total": 50,
              },
            ],
            "datePosted": "11/01/2017",
            "details": "We used a lot of water.",
            "duedate": "11/9/17",
            "id": "b3019762705779",
            "parsedDuedate": 1510210800000,
            "postedBy": Object {
              "house": "Dog9",
              "id": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
              "name": "Amanda Brenner",
            },
            "title": "Water Bill",
            "total": "100",
          },
        ],
        "bulletins": Array [
          Object {
            "datePosted": "11/01/2017",
            "details": "The dogs got out through a hole in the fence yesterday.",
            "hasRead": Array [
              "RVde1m9BhkQBLZlTJNsFlQZXqP43",
            ],
            "id": 1509567098747,
            "postedBy": Object {
              "house": "Dog9",
              "id": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
              "name": "Amanda Brenner",
            },
            "title": "Don't leave dogs out unattended.",
          },
        ],
        "chores": Array [
          Object {
            "assignedTo": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
            "datePosted": "10/31/2017",
            "details": "Empty the dishwasher.",
            "done": false,
            "id": "c1509485241076",
            "postedBy": Object {
              "house": "Dog9",
              "id": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
              "name": "Amanda Brenner",
            },
            "title": "Dishes",
            "urgency": "low",
          },
        ],
        "houseCode": "woof9",
        "houseKey": "-KxoIBWy5UYRscFYdWty",
        "houseName": "Dog9",
        "users": Array [
          Object {
            "house": "Dog9",
            "id": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
            "name": "Amanda Brenner",
          },
          Object {
            "house": "Dog9",
            "id": "cFNdorl5ceRyMr9XGvPFRVEcw7I3",
            "name": "Lola Koala",
          },
        ],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <p
          className="title"
        >
          Chore: Dishes
        </p>,
        <p
          className="urgency"
        >
          Urgency: low
        </p>,
        <p
          className="posted-by-on"
        >
          Posted by Amanda Brenner on 10/31/2017
        </p>,
        <p
          className="claim-chore"
          onClick={[Function]}
        >
          Claimed by Amanda Brenner
        </p>,
        <p
          className="details"
        >
          Details: Empty the dishwasher.
        </p>,
        <p
          className="chore-done"
          onClick={[Function]}
        >
          Mark as Done
        </p>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": "Chore: Dishes",
          "className": "title",
        },
        "ref": null,
        "rendered": "Chore: Dishes",
        "type": "p",
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": "Urgency: low",
          "className": "urgency",
        },
        "ref": null,
        "rendered": "Urgency: low",
        "type": "p",
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": "Posted by Amanda Brenner on 10/31/2017",
          "className": "posted-by-on",
        },
        "ref": null,
        "rendered": "Posted by Amanda Brenner on 10/31/2017",
        "type": "p",
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": "Claimed by Amanda Brenner",
          "className": "claim-chore",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Claimed by Amanda Brenner",
        "type": "p",
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": "Details: Empty the dishwasher.",
          "className": "details",
        },
        "ref": null,
        "rendered": "Details: Empty the dishwasher.",
        "type": "p",
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": "Mark as Done",
          "className": "chore-done",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Mark as Done",
        "type": "p",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <p
            className="title"
          >
            Chore: Dishes
          </p>,
          <p
            className="urgency"
          >
            Urgency: low
          </p>,
          <p
            className="posted-by-on"
          >
            Posted by Amanda Brenner on 10/31/2017
          </p>,
          <p
            className="claim-chore"
            onClick={[Function]}
          >
            Claimed by Amanda Brenner
          </p>,
          <p
            className="details"
          >
            Details: Empty the dishwasher.
          </p>,
          <p
            className="chore-done"
            onClick={[Function]}
          >
            Mark as Done
          </p>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": "Chore: Dishes",
            "className": "title",
          },
          "ref": null,
          "rendered": "Chore: Dishes",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": "Urgency: low",
            "className": "urgency",
          },
          "ref": null,
          "rendered": "Urgency: low",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": "Posted by Amanda Brenner on 10/31/2017",
            "className": "posted-by-on",
          },
          "ref": null,
          "rendered": "Posted by Amanda Brenner on 10/31/2017",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": "Claimed by Amanda Brenner",
            "className": "claim-chore",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Claimed by Amanda Brenner",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": "Details: Empty the dishwasher.",
            "className": "details",
          },
          "ref": null,
          "rendered": "Details: Empty the dishwasher.",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": "Mark as Done",
            "className": "chore-done",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Mark as Done",
          "type": "p",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
